File: src/main/java/com/hileco/cortex/primitives/layer/StackLayer.java
package com.hileco.cortex.primitives.layer;

import java.util.Stack;

public class StackLayer<T> extends Layer<StackLayer<T>> {
    private final Stack<T> entries;
    private int length;

    public StackLayer(StackLayer<T> parent) {
        super(parent);
        this.entries = new Stack<>();
        this.length = parent == null ? 0 : parent.length;
    }

    public Stack<T> getEntries() {
        return entries;
    }

    public void setLength(int length) {
        this.length = length;
    }

    public int getLength() {
        return length;
    }

    @Override
    boolean isEmpty() {
        var parent = getParent();
        if (parent != null) {
            return entries.isEmpty() && parent.length != this.length;
        } else {
            return entries.isEmpty();
        }
    }
}

File: cortex-data/src/main/kotlin/com/hileco/cortex/data/layer/StackLayer.kt
package com.hileco.cortex.data.layer

class StackLayer<T>(parent: StackLayer<T>?) : Layer<StackLayer<T>>(parent) {
    val entries: MutableMap<Int, T> = HashMap()
    var length: Int = 0

    override val isEmpty: Boolean
        get() {
            return entries.isEmpty() && parent?.length == this.length
        }

    init {
        this.length = parent?.length ?: 0
    }
}
